workflow-bot-app/src/main/java/com/symphony/bdk/workflow/engine/WorkflowDirectGraphBuilder.java

```diff
@@ -0,0 +1,229 @@
+        }
+      }
+      directGraph.addChildTo(eventNodeId)
+          .gateway(onEvents.isExclusive() ? Gateway.EVENT_BASED : Gateway.PARALLEL)
+          .addChild(activityId);
+      directGraph.addParent(activityId, eventNodeId);
+    }
+  }
+
+  private Optional<String> retrieveCondition(ActivityCompletedEvent event, BaseActivity activity,
+      WorkflowDirectGraph directGraph) {
+    return event.getIfCondition() != null ? Optional.of(event.getIfCondition())
+        : Optional.ofNullable(activity.getIfCondition());
+  }
+
+  private void validateNodeId(String currentNodeId, String activityId) {
+    boolean activityUnknown = workflow.getActivities().stream()
+        .noneMatch(a -> a.getActivity().getId().equals(currentNodeId));
+    if (activityUnknown) {
+      throw new ActivityNotFoundException(workflow.getId(), currentNodeId, activityId);
+    }
+  }
+
+  private void computeSignal(WorkflowDirectGraph directGraph, Event event, String eventNodeId, int activityIndex,
+      String activityId, List<Activity> activities) {
+    WorkflowNode signalEvent = new WorkflowNode().id(eventNodeId).event(event);
+    if (isFormRepliedEvent(event)) {
+      validateNodeId(eventNodeId.substring(WorkflowEventToCamundaEvent.FORM_REPLY_PREFIX.length()), activityId);
+      if (event instanceof EventWithTimeout && StringUtils.isEmpty(((EventWithTimeout) event).getTimeout())) {
+        ((EventWithTimeout) event).setTimeout(DEFAULT_FORM_REPLIED_EVENT_TIMEOUT);
+      }
+      directGraph.registerToDictionary(eventNodeId, signalEvent.elementType(WorkflowNodeType.FORM_REPLIED_EVENT));
+    } else {
+      Activity activity = activities.get(activityIndex);
+      String timeout = activity.getActivity().getOn().getTimeout(); // timeout value from on event
+      if ((event instanceof EventWithTimeout && StringUtils.isNotEmpty(((EventWithTimeout) event).getTimeout()))
+          // timeout value from activity itself
+          || StringUtils.isNotEmpty(timeout)) {
+        timeout = timeout == null ? ((EventWithTimeout) event).getTimeout() : timeout;
+        // in case of oneOf events list, there might be a timeout setup at 'on' level, this timeout will be applied to
+        // every event in the onfOf list, check if it is already registered, avoid repeating multiple of them
+        String newTimeoutEventId = eventNodeId + TIMEOUT_SUFFIX;
+        String parentId = directGraph.getParents(eventNodeId).get(0);
+        registerTimeoutEvent(directGraph, newTimeoutEventId, parentId, timeout);
+      }
+      directGraph.registerToDictionary(eventNodeId, signalEvent.elementType(WorkflowNodeType.SIGNAL_EVENT));
+    }
+  }
+
+  private String computeExpiredActivity(Event event, String activityId, WorkflowDirectGraph directGraph) {
+    // get the parent timeout event of the referred activity
+    String parentActivity = directGraph.getParents(event.getActivityExpired().getActivityId()).get(0);
```

Comment:
Are we sure the current activity, has a parent? An activity with activity-expired event should not be the first one of the workflow, but do we have something checking this constraint in the refactored code? Otherwise, this line can eventually  throw a NP.